rasa/shared/nlu/training_data/features.py

```diff
@@ -136,3 +158,224 @@ def fingerprint(self) -> Text:
         return rasa.shared.utils.io.deep_container_fingerprint(
             [self.type, self.origin, self.attribute, f_as_text]
         )
+
+    @staticmethod
+    def filter(
+        features_list: List[Features],
+        attributes: Optional[Iterable[Text]] = None,
+        type: Optional[Text] = None,
+        origin: Optional[List[Text]] = None,
+        is_sparse: Optional[bool] = None,
+    ) -> List[Features]:
+        """Filters the given list of features.
+
+        Args:
+          features_list: list of features to be filtered
+          attributes: List of attributes that we're interested in. Set this to `None`
+            to disable this filter.
+          type: The type of feature we're interested in. Set this to `None`
+            to disable this filter.
+          origin: If specified, this method will check that the exact order of origins
+            matches the given list of origins. The reason for this is that if
+            multiple origins are listed for a Feature, this means that this feature
+            has been created by concatenating Features from the listed origins in
+            that particular order.
+          is_sparse: Defines whether all features that we're interested in should be
+            sparse. Set this to `None` to disable this filter.
+
+        Returns:
+            sub-list of features with the desired properties
+        """
+        filtered = features_list
+        if attributes is not None:
+            attributes = set(attributes)
+            filtered = [f for f in filtered if f.attribute in attributes]
+        if origin is not None:
+            filtered = [
+                f
+                for f in filtered
+                if (f.origin if not isinstance(f.origin, Text) else list([f.origin]))
+                == origin
+            ]
+        if type is not None:
+            filtered = [f for f in filtered if f.type == type]
+        if is_sparse is not None:
+            filtered = [f for f in filtered if f.is_sparse() == is_sparse]
```

Comment:
If I can build up the lambda so that there are no `if <xy> is None` in it, then I think that would be nicer 🤔  (just wanted to avoid having four ifs for every feature in the list)

Reply:
mhm, maybe via a loop?

Reply:
If we use a loop then we have the complex control flow (i.e. 4 ifs) for every item in the list, which also doesn't look so nice, no? Could do something like...
```
filter = lambda x : True
if <xy> is not None:
    filter = lambda x:  filter(x) and (<evaluate condition wrt xy>)
...
filtered = [f for f in features_list_whatever if filter(f)]
```
🤔